parametric/apps/golang/otel_methods.go

```diff
@@ -0,0 +1,175 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	ot_api "go.opentelemetry.io/otel/trace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+func (s *apmClientServer) OtelStartSpan(ctx context.Context, args *OtelStartSpanArgs) (*OtelStartSpanReturn, error) {
+	fmt.Println("started_StartOtelSpan")
+
+	var pCtx context.Context = context.Background()
+	if pid := args.GetParentId(); pid != "" {
+		parent, ok := s.otelSpans[pid]
+		if ok {
+			pCtx = tracer.ContextWithSpan(context.Background(), parent.(ddtrace.Span))
+		}
+	}
+	var otelOpts = []ot_api.SpanStartOption{
+		ot_api.WithSpanKind(ot_api.ValidateSpanKind(ot_api.SpanKind(args.GetSpanKind()))),
+	}
+	if args.GetNewRoot() {
+		otelOpts = append(otelOpts, ot_api.WithNewRoot())
+	}
+	if t := args.GetTimestamp(); t != 0 {
+		tm := time.Unix(t, 0)
+		otelOpts = append(otelOpts, ot_api.WithTimestamp(tm))
+	}
+
+	// if attrs := args.GetAttributes(); attrs != nil {
+	// 	for k, v := range attrs {
+	// 		otelOpts = append(otelOpts, ot_api.WithAttributes(attribute.String(k, v)))
+	// 	}
+	// }
+
+	_, span := s.tp.Tracer("").Start(pCtx, args.Name, otelOpts...)
+	s.otelSpans[span.SpanContext().SpanID().String()] = span
+	return &OtelStartSpanReturn{
+		SpanId: span.SpanContext().SpanID().String(),
+	}, nil
+}
+
+func (s *apmClientServer) OtelEndSpan(ctx context.Context, args *OtelEndSpanArgs) (*OtelEndSpanReturn, error) {
+	span, ok := s.otelSpans[args.Id]
+	if !ok {
+		fmt.Sprintf("EndOtelSpan call failed, span with id=%s not found", args.Id)
+	}
+	// todo pass end span options
+	span.End()
+
+	return &OtelEndSpanReturn{}, nil
+}
+
+func (s *apmClientServer) OtelSetAttributes(ctx context.Context, args *OtelSetAttributesArgs) (*OtelSetAttributesReturn, error) {
+	span, ok := s.otelSpans[args.SpanId]
+	if !ok {
+		fmt.Sprintf("SetAttributes call failed, span with id=%s not found", args.SpanId)
+	}
+	for k, lv := range args.Attributes.KeyVals {
+		n := len(lv.GetVal())
+		if n == 0 {
+			continue
+		}
+		// all values are represented as slices
+		first := lv.GetVal()[0]
+		switch first.Val.(type) {
+		case *AttrVal_StringVal:
+			inp := make([]string, n)
+			for _, v := range lv.GetVal() {
+				inp = append(inp, v.GetStringVal())
+			}
+			span.SetAttributes(attribute.StringSlice(k, inp))
```

Comment:
The most common scenario would be a singular bool, int, etc.
```suggestion
			inp := make([]string, n)
			for _, v := range lv.GetVal() {
				inp = append(inp, v.GetStringVal())
			}
			if len(inp) == 1{
			span.SetAttributes(attribute.String(k, inp[0]))
			}
			span.SetAttributes(attribute.StringSlice(k, inp))
```